pkg/interop/client.go

```diff
@@ -0,0 +1,506 @@
+	GetAppSKey(ctx context.Context, asID string, req *ttnpb.SessionKeyRequest) (*ttnpb.AppSKeyResponse, error)
+}
+
+type prefixJoinServerClient struct {
+	joinServerClient
+	prefix types.EUI64Prefix
+}
+
+type Client struct {
+	joinServers []prefixJoinServerClient // Sorted by JoinEUI prefix range length.
+}
+
+var errUnknownProtocol = errors.DefineInvalidArgument("unknown_protocol", "unknown protocol")
+
+// ClientConfig represents the client-side interoperability through LoRaWAN Backend Interfaces configuration.
+type ClientConfig struct {
+	ConfigURI string        `name:"config-uri" description:"URI of the configuration file"`
+	CacheTime time.Duration `name:"cache-time"`
+}
+
+var errUnsupportedURIScheme = errors.DefineUnimplemented("unsupported_uri_scheme", "URI scheme `{scheme}` is not supported", "scheme")
+
+type tlsConfig struct {
+	RootCA      string `yaml:"root-ca"`
+	Certificate string `yaml:"certificate"`
+	Key         string `yaml:"key"`
+}
+
+func (conf tlsConfig) IsZero() bool {
+	return conf == (tlsConfig{})
+}
+
+func openFile(fs afero.Fs, basePath, path string) (afero.File, error) {
+	if !filepath.IsAbs(path) {
+		path = filepath.Join(basePath, path)
+	}
+	return fs.Open(path)
+}
+
+func readFile(fs afero.Fs, basePath, path string) ([]byte, error) {
+	f, err := openFile(fs, basePath, path)
+	if err != nil {
+		return nil, err
+	}
+	return ioutil.ReadAll(f)
+}
+
+func (conf tlsConfig) TLSConfig(fs afero.Fs, confDir string) (*tls.Config, error) {
+	var rootCAs *x509.CertPool
+	if conf.RootCA != "" {
+		caFile, err := openFile(fs, confDir, conf.RootCA)
+		if err != nil {
+			return nil, err
+		}
+
+		caPEM, err := ioutil.ReadAll(caFile)
+		if err != nil {
+			return nil, err
+		}
+		rootCAs = x509.NewCertPool()
+		rootCAs.AppendCertsFromPEM(caPEM)
+	}
+
+	var getCert func(*tls.CertificateRequestInfo) (*tls.Certificate, error)
+	if conf.Certificate != "" || conf.Key != "" {
+		getCert = func(*tls.CertificateRequestInfo) (*tls.Certificate, error) {
+			certFile, err := openFile(fs, confDir, conf.Certificate)
+			if err != nil {
+				return nil, err
+			}
+			certPEM, err := ioutil.ReadAll(certFile)
+			if err != nil {
+				return nil, err
+			}
+
+			keyFile, err := openFile(fs, confDir, conf.Key)
+			if err != nil {
+				return nil, err
+			}
+			keyPEM, err := ioutil.ReadAll(keyFile)
+			if err != nil {
+				return nil, err
+			}
+
+			cert, err := tls.X509KeyPair(certPEM, keyPEM)
+			if err != nil {
+				return nil, err
+			}
+			return &cert, nil
+		}
+	}
+	return &tls.Config{
+		RootCAs:              rootCAs,
+		GetClientCertificate: getCert,
+	}, nil
+}
+
+func decodeYAML(r io.Reader, v interface{}) error {
+	dec := yaml.NewDecoder(r)
+	dec.SetStrict(true)
+	return dec.Decode(v)
+}
+
+func NewClient(ctx context.Context, conf ClientConfig, fallbackTLS *tls.Config) (*Client, error) {
+	u, err := url.Parse(conf.ConfigURI)
```

Comment:
@htdvisser I meant usage of Afero here.

I'm fine with using Afero _in_ `pkg/fetch`, but now we get:

- Configuration of directory or URL through config and usage of `pkg/fetch` in `pkg/frequencyplans` and `pkg/devicerepository` and `pkg/applicationserver/io/web` (for templates)
- A URL  in `pkg/interop` which can start with `file` for files and anything else for URL

I don't like that inconsistency

Reply:
I agree

Reply:
So what is the proposal?
The deadline for this is tomorrow *and* (as far as I'm concerned) it works.
Can we open a technical debt issue about this?

Reply:
No, because fixing it breaks configuration and we cannot do that, see our compatibility commitment.

The proposal is to be consistent with what we have in;
- `pkg/frequencyplans`
- `pkg/devicerepository`
- `pkg/applicationserver/io/web`

I.e. config strings for folders vs URL, and use `pkg/fetch`.

Reply:
This is not addressed still